src/tlo/methods/equipment.py

```diff
@@ -0,0 +1,254 @@
+        self._all_item_descriptors = set(self._item_code_lookup.keys())
+        self._all_item_codes = set(self._item_code_lookup.values())
+
+        # Initialise the internal stores of the probability with which equipment items that are available.
+        self._set_equipment_items_available(availability=availability)
+
+    def on_simulation_end(self):
+        """Things to do when the simulation ends:
+         * Log (to the summary logger) the equipment that has been used.
+        """
+        self.write_to_log()
+
+    def update_availability(self, availability: str) -> None:
+        """Update the availability of equipment. This is expected to be called midway through the simulation if
+        the assumption of the equipment availability is changed."""
+        self._set_equipment_items_available(availability=availability)
+
+    def _set_equipment_items_available(self, availability: str):
+        """Update internal store of probabilities of items of equipment being available. This is called at the beginning
+         of the simulation and whenever an update in `availability` is done by `update_availability`."""
+
+        # All facility_id in the simulation
+        all_fac_ids = self.master_facilities_list['Facility_ID'].unique()
+
+        # All equipment items in the catalogue
+        all_eq_items = self.catalogue["Item_Code"].unique()
+
+        # Create "full" dataset, where we force that there is probability of availability for every item_code at every
+        # observed facility
+        df = pd.Series(
+            index=pd.MultiIndex.from_product(
+                [all_fac_ids, all_eq_items], names=["Facility_ID", "Item_Code"]
+            ),
+            data=float("nan"),
+        ).combine_first(
+            self.data_availability.set_index(["Facility_ID", "Item_Code"])[
+                "Pr_Available"
+            ]
+        )
+
+        # Merge in original dataset and use the mean in that facility_id to impute availability of missing item_codes
+        df = df.groupby("Facility_ID").transform(lambda x: x.fillna(x.mean()))
+        # ... and also impute availability for any facility_ids for which no data, based on all other facilities
+        df = df.groupby("Item_Code").transform(lambda x: x.fillna(x.mean()))
+
+        # Check no missing values
+        assert not df.isnull().any()
+
+        # Over-write these data if `availability` argument specifies that `none` or `all` items should be available
+        if availability == "default":
+            pass
+        elif availability == "all":
+            df = (df + 1).clip(upper=1.0)  # All probabilities -> 1.0
+        elif availability == "none":
+            df = df.mul(0.0)  # All probabilities -> 0.0
+        else:
+            raise KeyError(f"Unknown equipment availability specified: {availability=}")
+
+        # Save
+        self._probabilities_of_items_available = df
```

Comment:
Suggested changes in https://github.com/UCL/TLOmodel/pull/1342.

It looks like every time the equipment availability is updated, we are going to rebuild this dataframe from scratch which would likely slow us down a bit. 

In two of the cases ("all" and "none") we actually just want all the probabilities to be 1/0 respectively, and in the third case we can get away with computing the probabilities once (say at the start of the simulation) and then storing them rather than overwriting them with 1s and 0s.

So I think it might be better for us to store the "availability" (that gets passed on `__init__`) and slightly rework the probability method to check against this variable before trying to lookup equipment:

```python

def __init__(self, ...):
  # ...
  self.availabilty = "all"/"none"/"default"
  self._equip_probabilities_by_default = _set_equipment_items_available()
  # ...

def probability_all_equipment_available(self, ...):
  if self.availability = "all":
    return 1
  elif self.availability = "none":
    return 0
  else:
    return p # the value we currently compute, from looking up all availability probabilities and multiplying, etc.
```

We might also save a little bit of time in "all"/"none" by not having to query the dataframe (if it is big) too.